substrate/runtime/contract/src/vm/prepare.rs

```diff
@@ -0,0 +1,285 @@
+		// Inspect the module to extract the initial and maximum page count.
+		let limits = memory_type.limits();
+		match (limits.initial(), limits.maximum()) {
+			(initial, Some(maximum)) if initial > maximum => {
+				// Requested initial number of pages should not exceed the requested maximum.
+				return Err(Error::Memory);
+			}
+			(_, Some(maximum)) if maximum > config.max_memory_pages => {
+				// Maximum number of pages should not exceed the configured maximum.
+				return Err(Error::Memory);
+			}
+			(_, None) => {
+				// Maximum number of pages should be always declared.
+				// This isn't a hard requirement and can be treated as a maxiumum set
+				// to configured maximum.
+				return Err(Error::Memory);
+			}
+			(initial, maximum) => sandbox::Memory::new(initial, maximum),
+		}
+	} else {
+		// If none memory imported then just crate an empty placeholder.
+		// Any access to it will lead to out of bounds trap.
+		sandbox::Memory::new(0, Some(0))
+	};
+	let memory = memory.map_err(|_| Error::Memory)?;
+
+	Ok(PreparedContract {
+		instrumented_code: contract_module.into_wasm_code()?,
+		memory,
+	})
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use std::fmt;
+	use tests::Test;
+	use vm::tests::MockExt;
+	use wabt;
+
+	impl fmt::Debug for PreparedContract {
+		fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+			write!(f, "PreparedContract {{ .. }}")
+		}
+	}
+
+	fn parse_and_prepare_wat(wat: &str) -> Result<PreparedContract, Error> {
+		let wasm = wabt::Wat2Wasm::new().validate(false).convert(wat).unwrap();
+		let config = Config::<Test>::default();
+		let env = ::vm::env_def::init_env();
+		prepare_contract::<MockExt>(wasm.as_ref(), &config, &env)
+	}
+
+	#[test]
+	fn internal_memory_declaration() {
+		let r = parse_and_prepare_wat(r#"(module (memory 1 1))"#);
+		assert_matches!(r, Err(Error::InternalMemoryDeclared));
+	}
+
+	#[test]
+	fn memory() {
+		// This test assumes that maximum page number is configured to a certain number.
+		assert_eq!(Config::<Test>::default().max_memory_pages, 16);
+
+		let r = parse_and_prepare_wat(r#"(module (import "env" "memory" (memory 1 1)))"#);
+		assert_matches!(r, Ok(_));
+
+		// No memory import
+		let r = parse_and_prepare_wat(r#"(module)"#);
+		assert_matches!(r, Ok(_));
+
+		// initial exceed maximum
+		let r = parse_and_prepare_wat(r#"(module (import "env" "memory" (memory 16 1)))"#);
+		assert_matches!(r, Err(Error::Memory));
+
+		// no maximum
+		let r = parse_and_prepare_wat(r#"(module (import "env" "memory" (memory 1)))"#);
+		assert_matches!(r, Err(Error::Memory));
+
+		// requested maximum exceed configured maximum
+		let r = parse_and_prepare_wat(r#"(module (import "env" "memory" (memory 1 17)))"#);
+		assert_matches!(r, Err(Error::Memory));
+	}
+
+	#[test]
+	fn imports() {
+		// nothing can be imported from non-"env" module for now.
+		let r = parse_and_prepare_wat(r#"(module (import "vne" "memory" (memory 1 1)))"#);
```

Comment:
`"vne"` looks too much like an actual typo - maybe rename to `"another_module"`

Reply:
Agreed!